RuddockWebsite/auth_utils.py

```diff
@@ -0,0 +1,291 @@
+    if len(algorithms) == 0 or \
+        len(algorithms) != len(rounds) or len(algorithms) != len(salts):
+      return False
+
+    # Update with parsed values.
+    self.algorithms = algorithms
+    self.rounds = rounds
+    self.salts = salts
+    self.password_hash = password_hash
+    return True
+
+  def verify_password(self, password):
+    '''
+    Verifies a password by applying each algorithm in turn to the password.
+    Returns True if successful, else False.
+    '''
+    if self.password_hash is None:
+      # Not initialized properly.
+      return False
+
+    test_hash = password
+    for i in xrange(len(self.algorithms)):
+      algorithm = self.algorithms[i]
+      rounds = self.rounds[i]
+      salt = self.salts[i]
+      test_hash = hash_password(test_hash, salt, rounds, algorithm)
+    return test_hash == self.password_hash
+
+  def is_legacy(self):
+    ''' Returns true if the hashing algorithm is not the most current version. '''
+    return len(self.algorithms) != 1 or \
+        self.algorithms[0] != constants.PWD_HASH_ALGORITHM or \
+        self.rounds[0] != constants.HASH_ROUNDS
+
+def hash_password(password, salt, rounds, algorithm):
+  '''
+  Hashes the password with the salt and algorithm provided. The supported
+  algorithms are in PasswordHashParser.valid_algorithms.
+
+  Returns just the hash (not the full hash string). Returns None if an error
+  occurs.
+
+  Algorithms using the passlib library are returned in base64 format.
+  Algorithms using the hashlib library are returned in hex format.
+  '''
+  if algorithm == 'pbkdf2_sha256':
+    # Rounds must be set.
+    if rounds is None:
+      return None
+    result = passlib.hash.pbkdf2_sha256.encrypt(password, salt=salt, rounds=rounds)
+    # Return just the hash.
+    return result.split('$')[-1]
+  elif algorithm == 'md5':
+    # Rounds is ignored.
+    return hashlib.md5(salt + password).hexdigest()
+  return None
+
+def set_password(username, password):
+  ''' Sets the user's password. Automatically generates a new salt. '''
+  algorithm = constants.PWD_HASH_ALGORITHM
+  rounds = constants.HASH_ROUNDS
+  salt = generate_salt()
+  password_hash = hash_password(password, salt, rounds, algorithm)
+
+  # Create new password hash string.
+  parser = PasswordHashParser([algorithm], [rounds], [salt], password_hash)
+  full_hash = str(parser)
+  query = text("UPDATE users SET password_hash=:ph WHERE username=:u")
+  g.db.execute(query, ph=full_hash, u=username)
+  return
+
+def generate_salt():
+  ''' Generates a pseudorandom salt. '''
+  return misc_utils.generate_random_string(constants.SALT_SIZE)
+
+def generate_reset_key():
+  '''
+  Generates a pseudorandom reset key. Guaranteed to be unique in the database.
+  We use only digits and lowercase letters since the database string comparison
+  is case insensitive (we already have more than enough entropy anyway).
+  '''
+  chars = string.ascii_lowercase + string.digits
+  while True:
```

Comment:
No, we shouldn't try this more than once. If there's any collision, there's something very bad going on with the random number generator, devs should be notified, and we should bail out.